[FRONTEND] Shopping cart page master component added.

abdullahyldz opened the pull request:
## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Description

Shopping cart page master component is created. It will consist of ShoppingCartItems component when @ozdenizdolu creates it. 
This component send current items and their amounts to left component as props. Also, sends onChangeCart function as props to listen to changes in the left part. Then, updates the total price.


# Screenshots

![image](https://user-images.githubusercontent.com/44057640/99733406-2cea8d80-2ad2-11eb-84ef-cdbbe7067428.png)

# Test results

Please run all the tests and dump the result/summary here in a code block.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have checked my code and corrected any misspellings
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Potential problems/Important remarks

Values of shopping cart, can be obtained from AppContext.js
Layout problems exist.

Commit:
Shopping cart page master component added.

ozdenizdolu commented on ecommerce/frontend/src/components/ShoppingCart/ShoppingCartMaster.js:
```diff
@@ -0,0 +1,67 @@
+import { Button } from 'antd'
+import { Statistic, Row, Col } from 'antd'
+
+export const ShoppingCart = ({ currency = 'TL', price = 5, cart }) => {
```
In my opinion, It would be better to get products in the cart as a prop here. Then we can make a request to the backend in the `ShoppingCartPage.js` and send it to this component as a prop. It's just a design choice though.

ozdenizdolu replied:
Now that I think, If we are going to show some type loading icon, it might be better to do the request here and pull the products from inside this component.

ozdenizdolu reviewed (dismissed):
I have some comments on aesthetics below, feel free to ignore them:

- I think there is no need for its name to end with Master. 
- I think the black borders look bold, they are kinda killing the vibe. I don't have a better suggestion though :)

I think we cannot merge it as it is because there should be some mockup (but not empty) products that have been sent to ShoppingCartItems to show them. (After I implemented it of course)

ozdenizdolu reviewed (commented):
-ignore, this was a duplicate comment-

Commit:
husky test

abdullahyldz commented:
> I have some comments on aesthetics below, feel free to ignore them:
> 
> * I think there is no need for its name to end with Master.
> * I think the black borders look bold, they are kinda killing the vibe. I don't have a better suggestion though :)
> 
> I think we cannot merge it as it is because there should be some mockup (but not empty) products that have been sent to ShoppingCartItems to show them. (After I implemented it of course)

All agreed.

furkan-varol commented on ecommerce/frontend/src/components/ShoppingCart/ShoppingCartMaster.js:
```diff
@@ -0,0 +1,68 @@
+        setItemCount(newCount)
+    }
+
+    // husky test change
```
irrelevant comment

Commit:
useEffect, props change

MehdiSaffar commented on ecommerce/frontend/src/components/ShoppingCart/ShoppingCart.js:
```diff
@@ -0,0 +1,88 @@
+
+        onChangeCart(cart)
+        setIsLoading(false)
+    })
```
It is dangerous to forget the empty dependencies array as a second argument to useEffect as this will make it run everytime the component renders
```js
useEffect(() => {....}, [])
```

MehdiSaffar replied:
I think you shouldn't be calling onChangeCart here
also you are already doing setIsLoading(false) inside the fetch function itself. Please remove this one.

MehdiSaffar commented on ecommerce/frontend/src/components/ShoppingCart/ShoppingCart.js:
```diff
@@ -0,0 +1,88 @@
+                notification.error({ description: 'Failed to get shopping cart from backend' })
+            } finally {
+                setIsLoading(false)
+            }
+            fetch(`localhost:4000/user/${'userId'}/shoppingCart`).then(res => onChangeCart(res))
+        }
```
We will be using axios instead of fetch as it is more developer-friendly. It will also allow us to provide a base url which is important when moving from local testing to production.

MehdiSaffar replied:
Also when using inline formatting you should pass the variable directly, without using quotes like `${userId}`

MehdiSaffar replied:
Oh and that fetch should actually be inside the try { }, no?
And did you try it? because I think fetch is calling itself recursively forever?

MehdiSaffar commented on ecommerce/frontend/src/components/ShoppingCart/ShoppingCart.js:
```diff
@@ -0,0 +1,88 @@
+        } else {
+            alert('Total price: ' + totalPrice)
+        }
+    }
```
it would be better to use Antd's notification or modal dialogs (they have an imperative interface similar to `alert`). 
Maybe disable the checkout button when shopping cart is empty? or have a whole different look to the page using Antd's `<Empty/>` component?

MehdiSaffar commented:
I agree with @ozdenizdolu's comment about the look of the borders. It is better not use borders or at most a very faint color

Commit:
Created the ShoppingCartItem and ShoppingCartItems components

Commit:
Prepare both of the components for merging with main wrapper

ShoppingCartItem and ShoppingCartItems have been prepared for getting put into the layout

Commit:
fixed the conflicts between F-shopping-cart-master and F-shopping-cart-left-panel

ozdenizdolu commented:
I have merged my components into this one.

Commit:
package.json merge conflict solved

Commit:
Shopping cart integration is complete and tested manually

MehdiSaffar commented:
fixed key and margin-bottom errors

MehdiSaffar committed:
fix key

MehdiSaffar committed:
fix margin-bottom

MehdiSaffar committed:
add link to product in title

MehdiSaffar committed:
add vendor link correctly

MehdiSaffar committed:
A lot of refactoring, functions in AppContext

Commit:
merge master for axios changes -> solve conflicts

Commit:
Merge remote-tracking branch 'origin/master' into F-shopping-cart-master

abdullahyldz committed:
input number onChange changed to onPressEnter

abdullahyldz committed:
NOT TESTED: get shopping cart query

abdullahyldz committed:
Merge branch 'master' into F-shopping-cart-master

abdullahyldz committed:
get shopping cart, tested but cors policy problem

abdullahyldz committed:
Merge remote-tracking branch 'origin/master' into F-shopping-cart-master

abdullahyldz committed:
post request add shopping cart item works

abdullahyldz committed:
shopping cart changes

MehdiSaffar committed:
Merge remote-tracking branch 'origin/master' into F-shopping-cart-master

MehdiSaffar committed:
harmless fixes

MehdiSaffar committed:
improve design

abdullahyldz committed:
shopping cart -> bug fix , filter in getshoppingcart for 0 amount, delete bug fix

MehdiSaffar committed:
Multiple changes

- Show real products in search results
- Change some CSS
- Show real image in shopping cart item
- Fix img contain
- Fix order of init and get shopping cart
- Set guest user as initial user
- Only fetch shopping cart when user type is customer
